File: utilities/src/main/java/xerus/util/helpers/Timer.java
package xerus.util.helpers;

/** times in milliseconds */
public class Timer {

	private static final boolean isNano = true;

	private static Timer t;
	private long time;
	
	public Timer() {
		restart();
	}
	
	public static void start() {
		t = new Timer();
	}

	public void restart() { time = System.currentTimeMillis(); }

	/** elapsed time in milliseconds */
	public static long runtime() {
		return t.time();
	}

	/** elapsed time in milliseconds */
	public long time() {
		return System.currentTimeMillis() - time;
	}
	
	public static long finish() {
		return finish("Time");
	}
	
	public static long finish(String msg) {
		if(t == null)
			return 0;
		long time = runtime();
		t = null;
		System.out.println(parseTime(msg, time));
		return t.printTime(msg);
	}

	public long printTime(String msg) {
		long time = time();
		System.out.println(parseTime(msg, time));
		return time;
	}
	
	public static String parseTime(String msg, long time) {
		StringBuilder res = new StringBuilder(msg).append(": ");
		if(isNano)
			time /= 1000000;
		if(time < 10000)
			res.append(time).append("m");
		else
			res.append((time/100) / 10.0);
		return res.append("s").toString();
	}

}

File: kotlin/src/xerus/ktutil/helpers/Timer.kt
package xerus.ktutil.helpers

import xerus.ktutil.formatTimeDynamic

/**
 * Records times in milliseconds.
 *
 * Can be used statically, which internally uses a static Timer instance,
 * or instantiated and then used.
 */
class Timer {
	
	private var time: Long = System.currentTimeMillis()
	
	/** resets the time to [System.currentTimeMillis] */
	fun restart() {
		time = System.currentTimeMillis()
	}
	
	/** elapsed time in milliseconds */
	fun time(): Long =
		System.currentTimeMillis() - time
	
	/**
	 * Pretty-prints the [time] using [parseTime] and returns it
	 * @param msg the message to prepend the time with or null for no message
	 * @return [time] */
	fun printTime(msg: String?): Long {
		val time = time()
		println(parseTime(msg, time))
		return time
	}
	
	companion object {
		
		private var t: Timer? = null
		
		/** Starts a new internal [Timer] */
		fun start() {
			t = Timer()
		}
		
		/** Elapsed time in milliseconds */
		fun runtime() = t!!.time()
		
		/** Pretty-prints the elapsed time using [parseTime] and resets the internal timer */
		fun finish(msg: String = "Time") =
			t?.run { printTime(msg).also { t = null } } ?: 0
		
		/**
		 * Returns the given [millis] with a message.
		 * If [millis] exceeds 10 seconds, it will show seconds with tens of seconds.
		 * If [millis] exceeds 300 seconds, it will show a digital clock-like representation.
		 *
		 * @param msg the message to prepend the time with or null for no message
		 * @param millis the milliseconds - will be postfixed by the appropriate unit */
		fun parseTime(msg: String?, millis: Long): String {
			val res = if(msg != null) StringBuilder(msg).append(": ") else StringBuilder()
			if(millis < 300_000) {
				if(millis < 10_000) {
					res.append(millis).append('m')
				} else {
					res.append(millis / 1000).append('.').append((millis / 100) % 10)
				}.append('s')
			} else {
				res.append(formatTimeDynamic(millis / 1000))
			}
			return res.toString()
		}
		
	}
	
}
